tinygraph/tinygraph.c

```diff
@@ -506,3 +507,334 @@ bool tinygraph_reorder(
   // allocation. For now this is always returning true.
   return true;
 }
+
+
+// The purpose of the dijkstra context is to cache state
+// like the distance and parents array, so that we don't
+// have to allocate memory for every new s-t search.
+typedef struct tinygraph_dijkstra {
+  uint32_t s;
+  uint32_t t;
+
+  uint32_t* dist;
+  uint32_t* parent;
+  tinygraph_array_s path;
+
+  const uint16_t* weight;
```

Comment:
One interesting optimization the size realm to look into is the following: `weight` is a bad way to store costs on a graph.

It is fast to have it simply as a number and that is why most people do this, but it is really wasteful if you want multiple metrics. Usually `distance` + `speed` is a better storage format. It has an additional advantage: In practical terms you can make `distance` an uint16 (e.g. ~65km max which is fine if you don't have CH-style shortcuts) and speed an uint8 (255 km/h). Every new metric will only need a different speed array.

Also speeds tend to be somewhat the same, in the same vicinity so that would make it even more compressible.

I'm wondering if there is a nice abstraction for this using some of the other primitives here.

Reply:
Good point! I started with a generic `uint16_t` weight to get something out and because so far we're not committed to the road network graph yet. The only road network related function in our API is the optional `reorder` function to re-order graph nodes based on a space-filling curve of their location.

At some point I believe we need to commit to one direction and stop keeping it as a generic graph library. At that point these optimizations will come into play.